module/src/main/scala/com/gu/googleauth/auth.scala

```diff
@@ -114,44 +181,43 @@ object GoogleAuth {
     discoveryDocument().map(dd => Redirect(s"${dd.authorization_endpoint}", queryString))
   }
 
-  def validatedUserIdentity(config: GoogleAuthConfig, expectedAntiForgeryToken: String)
+  def validatedUserIdentity(config: GoogleAuthConfig)
         (implicit request: RequestHeader, context: ExecutionContext, ws: WSClient): Future[UserIdentity] = {
-    if (!request.queryString.getOrElse("state", Nil).contains(expectedAntiForgeryToken)) {
-      throw new IllegalArgumentException("The anti forgery token did not match")
-    } else {
-      discoveryDocument().flatMap { dd =>
-        val code = request.queryString("code")
-        ws.url(dd.token_endpoint).post {
-          Map(
-            "code" -> code,
-            "client_id" -> Seq(config.clientId),
-            "client_secret" -> Seq(config.clientSecret),
-            "redirect_uri" -> Seq(config.redirectUrl),
-            "grant_type" -> Seq("authorization_code")
-          )
-        }.flatMap { response =>
-          googleResponse(response) { json =>
-              val token = Token.fromJson(json)
-              val jwt = token.jwt
-              config.domain foreach { domain =>
-                if (!jwt.claims.email.split("@").lastOption.exists(_ == domain))
-                  throw new GoogleAuthException("Configured Google domain does not match")
-              }
-              ws.url(dd.userinfo_endpoint)
-                .withHttpHeaders("Authorization" -> s"Bearer ${token.access_token}")
-                .get().map { response =>
-                  googleResponse(response) { json =>
-                      val userInfo = UserInfo.fromJson(json)
-                      UserIdentity(
-                        jwt.claims.sub,
-                        jwt.claims.email,
-                        userInfo.given_name,
-                        userInfo.family_name,
-                        jwt.claims.exp,
-                        userInfo.picture
-                      )
-                  }
-              }
+
+    config.antiForgeryChecker.verifyToken(request)
```

Comment:
This is more an improvement, but I find the handling of exceptions a bit confusing now.

Before your changes `validatedUserIdentity` was throwing only `IllegalArgumentException` if the validation failed, which was [specifically caught by the caller](https://github.com/guardian/play-googleauth/blob/master/module/src/main/scala/com/gu/googleauth/actions.scala#L129).

With your changes  `validatedUserIdentity` is calling `verifyToken` which can throw  various `jwt` exceptions (`SignatureException `, `ExpiredJwtException `, `UnsupportedJwtException`) so `validatedUserIdentity` will throw them as well, and it will be only caught by the [generic `Throwable: e`](https://github.com/guardian/play-googleauth/blob/master/module/src/main/scala/com/gu/googleauth/actions.scala#L133).      

I feel `verifyToken` should return a `Try[Unit]` and `validatedUserIdentity ` a `Future[UserIdentity]` and then pattern match on the type of exception in the caller (action) to display a useful message.

Reply:
Are the changes in 78eb60f ok for this? It looks like this now:

https://github.com/guardian/play-googleauth/blob/78eb60fe65cebbb9c8daddcf20633b724761e909/module/src/main/scala/com/gu/googleauth/auth.scala#L115-L129

 We're pretty keen on getting this released soon and I think the messages will be sufficiently clear for developers to interpret?

Reply:
👍